soroban-env-host/src/crypto/bls12_381.rs

```diff
@@ -0,0 +1,785 @@
+                    format!(
+                        "bls12-381 field element {}: invalid input length to deserialize",
+                        msg
+                    )
+                    .as_str(),
+                    &[
+                        Val::from_u32(bytes.len() as u32).into(),
+                        Val::from_u32(EXPECTED_SIZE as u32).into(),
+                    ],
+                ));
+            }
+            self.charge_budget(ContractCostType::MemCpy, Some(EXPECTED_SIZE as u64))?;
+            let mut buf = [0u8; EXPECTED_SIZE];
+            buf.copy_from_slice(bytes);
+            buf.reverse();
+            self.deserialize_uncompressed_no_validate::<EXPECTED_SIZE, _>(&buf, msg)
+        })
+    }
+
+    pub(crate) fn fp_deserialize_from_bytesobj(&self, bo: BytesObject) -> Result<Fq, HostError> {
+        // `CanonicalDeserialize for Fp<P, N>` assumes input bytes in
+        // little-endian order, with the highest bits being empty flags.
+        // thus we must first reverse the bytes before passing them in.
+        // there is no other check for Fp besides the length check.
+        self.field_element_deserialize::<FP_SERIALIZED_SIZE, Fq>(bo, "Fp")
+    }
+
+    pub(crate) fn fp2_deserialize_from_bytesobj(&self, bo: BytesObject) -> Result<Fq2, HostError> {
+        // `CanonicalDeserialize for QuadExtField<P>` reads the first chunk,
+        // deserialize it into `Fp` as c0. Then repeat for c1. The
+        // deserialization for `Fp` follows same rules as above, where the
+        // bytes are expected in little-endian, with the highest bits being
+        // empty flags. There is no check involved.
+        //
+        // This is entirely reversed from the [standard](https://github.com/zcash/librustzcash/blob/6e0364cd42a2b3d2b958a54771ef51a8db79dd29/pairing/src/bls12_381/README.md#serialization)
+        // we have adopted. This is the input format we provide:
+        //
+        // `input = be_bytes(c1) || be_bytes(c0)`
+        //
+        // So we just need to reverse our input.
+        self.field_element_deserialize::<FP2_SERIALIZED_SIZE, Fq2>(bo, "Fp2")
+    }
+
+    pub(crate) fn g1_vec_from_vecobj(&self, vp: VecObject) -> Result<Vec<G1Affine>, HostError> {
+        let len: u32 = self.vec_len(vp)?.into();
+        let mut points: Vec<G1Affine> = vec![];
+        self.charge_budget(
+            ContractCostType::MemAlloc,
+            Some(len as u64 * G1_SERIALIZED_SIZE as u64),
+        )?;
+        points.reserve(len as usize);
+        let _ = self.visit_obj(vp, |vp: &HostVec| {
+            for p in vp.iter() {
+                let pp = self.g1_affine_deserialize_from_bytesobj(
+                    BytesObject::try_from_val(self, p)?,
+                    true,
+                )?;
+                points.push(pp);
+            }
+            Ok(())
+        });
+        Ok(points)
+    }
+
+    pub(crate) fn scalar_vec_from_vecobj(&self, vs: VecObject) -> Result<Vec<Fr>, HostError> {
```

Comment:
nit: since this is a host method, I think it should be named less generically, like bls_scalar_... or fr_scalar_...

Reply:
done